File: keypad-view/src/main/java/com/maxpilotto/keypadview/Key.java
package com.maxpilotto.keypadview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.DrawableRes;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

public class Key extends LinearLayout {
    public static final float DEFAULT_TEXT_SIZE_SP = 20;

    private LinearLayout root;
    private TextView text;
    private ImageView icon;
    private int position;

    public Key(Context context) {
        this(context, null, 0);
    }

    public Key(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public Key(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        inflate(context, R.layout.key, this);

        root = findViewById(R.id.root);
        text = findViewById(R.id.text);
        icon = findViewById(R.id.icon);

        if (attrs != null) {
            TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.Key, defStyleAttr, 0);
            int txtSize = a.getDimensionPixelSize(R.styleable.Key_keyTextSize, 0);

            setText(a.getString(R.styleable.Key_keyValue));
            setTextColor(a.getColor(
                    R.styleable.Key_keyTextColor,
                    getResources().getColor(android.R.color.darker_gray)
            ));
            setKeyBackground(a.getResourceId(R.styleable.Key_keyBackground, R.drawable.key_background));

            if (txtSize != 0) {
                text.setTextSize(TypedValue.COMPLEX_UNIT_PX, txtSize);
            } else {
                text.setTextSize(DEFAULT_TEXT_SIZE_SP);
            }

            if (a.getDimensionPixelSize(R.styleable.Key_keyIconSize, 0) != 0) {
                setIconSize(a.getDimensionPixelSize(R.styleable.Key_keyIconSize, 0));
            }

            if (a.getResourceId(R.styleable.Key_keyWrapperBackground, 0) != 0) {
                setWrapperBackground(a.getResourceId(R.styleable.Key_keyWrapperBackground, 0));
            }

            if (a.getResourceId(R.styleable.Key_keyIcon, 0) != 0) {
                setIcon(a.getResourceId(R.styleable.Key_keyIcon, 0));
            }

            if (a.getColor(R.styleable.Key_keyTint, 0) != 0) {
                setIconTint(a.getColor(R.styleable.Key_keyTint, 0));
            }

            a.recycle();
        }
    }

    @Override public void setOnLongClickListener(View.OnLongClickListener l) {
        //NOTE: Without this the view won't be clickable

        root.setOnLongClickListener(l);
    }

    @Override public void setOnClickListener(View.OnClickListener l) {
        //NOTE: Without this the view won't be clickable

        root.setOnClickListener(l);
    }

    /**
     * Returns the position of the Key inside the Keypad, this can be used as a unique identifier
     *
     * @return Position of the Key inside the Keypad
     */
    public int getPosition() {
        return position;
    }

    /**
     * Sets the position of the Key, this will just set the position index and NOT change the layout
     *
     * @param position Position
     */
    public void setPosition(int position) {
        this.position = position;
    }

    /**
     * Sets the text of the key, this will hide the icon
     *
     * @param value Key's text
     */
    public void setText(String value) {
        hideIcon();

        text.setText(value);
    }

    /**
     * Returns the text of the key
     *
     * @return Key's text
     */
    public String getText() {
        return text.getText().toString();
    }

    /**
     * Sets the icon to this Key, this will hide the text
     *
     * @param iconRes Icon res
     */
    public void setIcon(@DrawableRes int iconRes) {
        hideText();

        icon.setImageResource(iconRes);
    }

    /**
     * Sets the tint to the icon
     */
    public void setIconTint(@ColorInt int color) {
        icon.setColorFilter(color, PorterDuff.Mode.SRC_IN);
    }

    /**
     * Sets the key background, by default it is a gray circle
     *
     * @param res Background
     */
    public void setKeyBackground(@DrawableRes int res) {
        text.setBackgroundResource(res);
        icon.setBackgroundResource(res);
    }

    /**
     * Returns the key background
     *
     * @return Background
     */
    public Drawable getKeyBackground() {
        return text.getBackground();
    }

    /**
     * Sets the key text color, gray by default
     *
     * @param color ColorInt
     */
    public void setTextColor(@ColorInt int color) {
        text.setTextColor(color);
    }

    /**
     * Returns the wrapper layout's background
     *
     * @return Background
     */
    public Drawable getWrapperBackground() {
        return root.getBackground();
    }

    /**
     * Sets the wrapper layout's background, this will disable the ripple effect
     * If you want to set your own background and keep the ripple effect, you have to implement the ripple animation by yourself
     *
     * @param res Background
     */
    public void setWrapperBackground(@DrawableRes int res) {
        root.setBackgroundResource(res);
    }

    /**
     * Sets the text size of the key
     *
     * @param sp Size in SP
     */
    public void setTextSize(float sp) {
        text.setTextSize(sp);
    }

    /**
     * Sets the sizes of the key, this will set both width and height
     *
     * @param size Sizes (width and height) in PX
     */
    public void setIconSize(int size) {
        LayoutParams params = (LayoutParams) icon.getLayoutParams();

        params.width = size;
        params.height = size;
    }

    /**
     * Returns the TextView that's used to display text, this is used for further customization
     *
     * @return Text's view
     */
    public TextView getTextView() {
        return text;
    }

    /**
     * Returns the ImageView that's used as icon, this is used for further customization
     *
     * @return Icon's view
     */
    public ImageView getImageView() {
        return icon;
    }

    /**
     * Replaces the text with an icon
     */
    public void hideIcon() {
        if (text.getVisibility() == GONE && icon.getVisibility() == VISIBLE) {
            text.setVisibility(VISIBLE);
            icon.setVisibility(GONE);
        }
    }

    /**
     * Replaces the icon with a text
     */
    public void hideText() {
        if (text.getVisibility() == VISIBLE && icon.getVisibility() == GONE) {
            text.setVisibility(GONE);
            icon.setVisibility(VISIBLE);
        }
    }
}

File: keypad-view/src/main/java/com/maxpilotto/keypadview/Key.kt
/*
 * Copyright 2018 Max Pilotto
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.maxpilotto.keypadview

import android.content.Context
import android.graphics.PorterDuff
import android.support.annotation.ColorInt
import android.support.annotation.DrawableRes
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import kotlinx.android.synthetic.main.key.view.*

/**
 * Representation of a single Keypad key
 *
 * Created on 07/08/2019 at 15:55
 */
class Key : LinearLayout {
    /**
     * Position of this Key inside a KeyPad, this could be used as a unique identifier
     */
    var position: Int

    /**
     * Key's text, this will hide the icon
     */
    var text: String
        set(value) {
            hideIcon()

            textView?.text = value
            field = value
        }

    /**
     * Key's text color
     */
    var textColor: Int
        set(@ColorInt value) {
            textView?.setTextColor(value)
            field = value
        }

    /**
     * Key's text size
     */
    var textSize: Float
        set(value) {
            textView?.textSize = value
            field = value
        }

    /**
     * Key's icon, this will hide the text
     */
    var icon: Int
        set(@DrawableRes value) {
            hideText()

            iconView?.setImageResource(value)
            field = value
        }

    /**
     * Key's icon size
     */
    var iconSize: Int
        set(value) {
            val params = iconView?.layoutParams

            params?.height = value
            params?.width = value
            field = value
        }

    /**
     * Key's icon tint
     */
    var iconTint: Int
        set(@ColorInt value) {
            iconView?.setColorFilter(value, PorterDuff.Mode.SRC_IN)
            field = value
        }

    /**
     * Key background
     */
    var background: Int
        set(@DrawableRes value) {
            textView?.setBackgroundResource(value)
            iconView?.setBackgroundResource(value)
            field = value
        }

    /**
     * Key wrapper background, this background won't overwrite the [background] and it should be used
     * to enable the ripple effect
     */
    var wrapperBackground: Int
        set(@DrawableRes value) {
            setBackgroundResource(value)
            field = value
        }

    init {
        position = 0
        text = ""
        textColor = 0
        textSize = DEFAULT_TEXT_SIZE
        icon = 0
        iconSize = 0
        iconTint = 0
        background = R.drawable.key_background
        wrapperBackground = 0
    }

    constructor(context: Context) : this(context, null, 0)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        View.inflate(context, R.layout.key, this)

        if (attrs != null) {
            val array = context.obtainStyledAttributes(attrs, R.styleable.Key, defStyleAttr, 0)
            val textSize = array.getDimensionPixelSize(R.styleable.Key_keyTextSize, 0)
            val icon = array.getResourceId(R.styleable.Key_keyIcon, 0)
            val iconSize = array.getDimensionPixelSize(R.styleable.Key_keyIconSize, 0)
            val tint = array.getColor(R.styleable.Key_keyTint, 0)
            val wrapper = array.getResourceId(R.styleable.Key_keyWrapperBackground, 0)

            text = array.getString(R.styleable.Key_keyValue)
            textColor = array.getColor(R.styleable.Key_keyTextColor, 0)
            background = array.getResourceId(R.styleable.Key_keyBackground, R.drawable.key_background)

            if (textSize != 0) {
                textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize.toFloat())
            }

            if (icon != 0) {
                this.icon = icon
            }

            if (iconSize != 0) {
                this.iconSize = iconSize
            }

            if (tint != 0) {
                this.iconTint = tint
            }

            if (wrapper != 0) {
                this.wrapperBackground = wrapper
            }

            array.recycle()
        }
    }

    override fun setOnClickListener(l: OnClickListener?) {
        root.setOnClickListener(l)
    }

    override fun setOnLongClickListener(l: OnLongClickListener?) {
        root.setOnLongClickListener(l)
    }

    /**
     * Returns the view used for the text, this can be used for further customization
     */
    fun getTextView(): TextView {
        return textView
    }

    /**
     * Returns the view used for the icon, this can be used for further customization
     */
    fun getIconView(): ImageView {
        return iconView
    }

    /**
     * Replaces the text with an icon
     */
    fun hideIcon() {
        if (textView?.visibility == View.GONE && iconView?.visibility == View.VISIBLE) {
            textView?.visibility = View.VISIBLE
            iconView?.visibility = View.GONE
        }
    }

    /**
     * Replaces the icon with a text
     */
    fun hideText() {
        if (textView?.visibility == View.VISIBLE && iconView?.visibility == View.GONE) {
            textView?.visibility = View.GONE
            iconView?.visibility = View.VISIBLE
        }
    }

    /**
     * Hides the key
     */
    fun hide() {
        visibility = View.INVISIBLE
    }

    /**
     * Shows the key
     */
    fun show() {
        visibility = View.VISIBLE
    }

    companion object {
        const val DEFAULT_TEXT_SIZE = 20F
    }
}